collect_app/src/test/java/org/odk/collect/android/instancemanagement/send/ReadyToSendViewModelTest.kt

```diff
@@ -0,0 +1,95 @@
+package org.odk.collect.android.instancemanagement.send
+
+import androidx.arch.core.executor.testing.InstantTaskExecutorRule
+import androidx.test.espresso.matcher.ViewMatchers.assertThat
+import org.hamcrest.Matchers.equalTo
+import org.junit.Rule
+import org.junit.Test
+import org.odk.collect.forms.instances.Instance
+import org.odk.collect.formstest.InMemInstancesRepository
+import org.odk.collect.shared.TimeInMs
+import org.odk.collect.testshared.FakeScheduler
+
+class ReadyToSendViewModelTest {
+    @get:Rule
+    val instantTaskExecutorRule = InstantTaskExecutorRule()
+
+    private val instancesRepository = InMemInstancesRepository().also {
+        it.save(
+            Instance.Builder()
+                .formId("1")
+                .status(Instance.STATUS_INCOMPLETE)
+                .lastStatusChangeDate(TimeInMs.ONE_SECOND * 1)
+                .build()
+        )
+
+        it.save(
+            Instance.Builder()
+                .formId("2")
+                .status(Instance.STATUS_COMPLETE)
+                .lastStatusChangeDate(TimeInMs.ONE_SECOND * 5)
+                .build()
+        )
+
+        it.save(
+            Instance.Builder()
+                .formId("3")
+                .status(Instance.STATUS_SUBMITTED)
+                .lastStatusChangeDate(TimeInMs.ONE_SECOND * 4)
+                .build()
+        )
+
+        it.save(
+            Instance.Builder()
+                .formId("4")
+                .status(Instance.STATUS_SUBMITTED)
+                .lastStatusChangeDate(TimeInMs.ONE_SECOND * 6)
+                .build()
+        )
+
+        it.save(
+            Instance.Builder()
+                .formId("5")
+                .status(Instance.STATUS_SUBMISSION_FAILED)
+                .lastStatusChangeDate(TimeInMs.ONE_SECOND * 7)
+                .build()
+        )
+
+        it.save(
+            Instance.Builder()
+                .formId("6")
+                .status(Instance.STATUS_COMPLETE)
+                .lastStatusChangeDate(TimeInMs.ONE_SECOND * 3)
+                .build()
+        )
+
+        it.save(
+            Instance.Builder()
+                .formId("7")
+                .status(Instance.STATUS_COMPLETE)
+                .lastStatusChangeDate(TimeInMs.ONE_SECOND * 10)
+                .build()
+        )
+    }
+    private val scheduler = FakeScheduler()
+
+    private val viewModel = ReadyToSendViewModel(instancesRepository, scheduler) { TimeInMs.ONE_SECOND * 10 }
+
+    @Test
+    fun `numberOfSentInstances should represent the real number of instances with STATUS_SUBMITTED in the database`() {
+        scheduler.runBackground()
```

Comment:
It surprised me that you didn't also have to call `runForeground` to have `data` updated by the foreground part of the task. It looks like `FakeScheduler` runs the foreground part for you though as it splits "background" and "foreground" on a slightly different level (the way the `Scheduler` was called really). Do you think we should change it so that `runForeground` would be required here? I could go either way on it really!

Reply:
The real implementation works in such a way that the foreground part is triggered automatically once the background one is done so I think in tests it can be like that too unless we need to do something in between. For now, we don't need that so I think we can leave it as-is.